src/shims/unix/fs.rs

```diff
@@ -1659,6 +1659,68 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             }
         }
     }
+
+    fn realpath(
+        &mut self,
+        path_op: &OpTy<'tcx, Tag>,
+        processed_path_op: &OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, Pointer<Option<Tag>>> {
+        let this = self.eval_context_mut();
+        this.assert_target_os_is_unix("realpath");
+
+        let pathname = this.read_path_from_c_str(this.read_pointer(path_op)?)?;
+
+        // Reject if isolation is enabled.
+        if let IsolatedOp::Reject(reject_with) = this.machine.isolated_op {
+            this.reject_in_isolation("`realpath`", reject_with)?;
+            let eacc = this.eval_libc("EACCES")?;
+            this.set_last_error(eacc)?;
+            return Ok(Pointer::null());
+        }
+
+        let result = std::fs::canonicalize(pathname);
+        match result {
+            Ok(resolved) => {
+                let processed_ptr = this.read_pointer(processed_path_op)?;
+                let dest = if this.ptr_is_null(processed_ptr)? {
+                    // "If resolved_path is specified as NULL, then realpath() uses
```

Comment:
This is a GNU extension. POSIX says

> If resolved_name is a null pointer, the behavior of realpath() is implementation-defined.

So if someone calls this on macOS, Miri might accept code that fails on real macOS.

Reply:
So, what does miri do in this case? Stick with posix? Add an opt-in for GNU? Would love pointers to code that handles a similar case.

FYI, it appears BSD behaves the same way w.r.t a null pointer: https://www.freebsd.org/cgi/man.cgi?query=realpath&sektion=3&format=html so perhaps it is a good default.

```c
char * realpath(const char * restrict pathname, char * restrict resolved_path);
```
> If the function was supplied NULL as resolved_path, and operation did	not cause errors, the returned value is a null-terminated string in a buffer allocated by a call to `malloc(3)`

Reply:
(FWIW this PR was developed on macos)

Reply:
is this a code path actually hit by libstd? Could we just turn it into an error path?

Reply:
Does miri support nostd? Also, we want to handle the cases where non-std crates use `libc`? If either of those are true I don't think we can make this case error

Reply:
Yeah, this is from the man page on macos:

>   As a permitted extension to the standard, if resolved_name is NULL,
     memory is allocated for the resulting absolute pathname, and is
     returned by realpath().  This memory should be freed by a call to
     free(3) when no longer needed.

Reply:
Everything I have checked does it, so let's default to doing it and see if anyone files a bug for other targets.